internal/querycoordv2/checkers/segment_checker.go

```diff
@@ -58,80 +59,164 @@ func (c *SegmentChecker) Check(ctx context.Context) []task.Task {
+	lacks, redundancies := c.getHistoricalSegmentDiff(c.targetMgr, c.dist, c.meta, replica.GetCollectionID(), replica.GetID())
 	tasks := c.createSegmentLoadTasks(ctx, lacks, replica)
 	ret = append(ret, tasks...)
 
 	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_All)
 	ret = append(ret, tasks...)
 
 	// compare inner dists to find repeated loaded segments
-	redundancies = findRepeatedSegments(dists)
+	redundancies = c.findRepeatedHistoricalSegments(c.dist, c.meta, replica.GetID())
 	redundancies = c.filterExistedOnLeader(replica, redundancies)
 	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_All)
 	ret = append(ret, tasks...)
 
-	// release redundant growing segments
-	leaderRedundancies := c.findNeedReleasedGrowingSegments(replica)
-	redundancies = make([]*meta.Segment, 0)
-	for _, segments := range leaderRedundancies {
-		redundancies = append(redundancies, segments...)
-	}
+	// compare with target to find the lack and redundancy of segments
+	_, redundancies = c.getStreamingSegmentDiff(c.targetMgr, c.dist, c.meta, replica.GetCollectionID(), replica.GetID())
 	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_Streaming)
 	ret = append(ret, tasks...)
 
 	return ret
 }
 
-func (c *SegmentChecker) getSegmentsDist(replica *meta.Replica) []*meta.Segment {
-	ret := make([]*meta.Segment, 0)
+// GetStreamingSegmentDiff get streaming segment diff between leader view and target
+func (c *SegmentChecker) getStreamingSegmentDiff(targetMgr *meta.TargetManager,
+	distMgr *meta.DistributionManager,
+	metaInfo *meta.Meta,
+	collectionID int64,
+	replicaID int64) (toLoad []*datapb.SegmentInfo, toRelease []*meta.Segment) {
+	replica := metaInfo.Get(replicaID)
+	if replica == nil {
+		log.Info("replica does not exist, skip it")
+		return
+	}
+	dist := c.getStreamingSegmentsDist(distMgr, replica)
+	distMap := typeutil.NewUniqueSet()
+	for _, s := range dist {
+		distMap.Insert(s.GetID())
+	}
+
+	nextTargetMap := targetMgr.GetStreamingSegmentsByCollection(collectionID, meta.NextTarget)
+	currentTargetMap := targetMgr.GetStreamingSegmentsByCollection(collectionID, meta.CurrentTarget)
+
+	//get segment which exist on next target, but not on dist
+	for segmentID, segment := range nextTargetMap {
+		if !distMap.Contain(segmentID) {
+			toLoad = append(toLoad, segment)
+		}
+	}
+
+	nextTargetChannelMap := targetMgr.GetDmChannelsByCollection(collectionID, meta.NextTarget)
+
+	// get segment which exist on dist, but not on current target and next target
+	for _, segment := range dist {
+		_, existOnCurrent := currentTargetMap[segment.GetID()]
+		_, existOnNext := nextTargetMap[segment.GetID()]
+
+		if !existOnNext && !existOnCurrent {
+			if channel, ok := nextTargetChannelMap[segment.InsertChannel]; ok {
+				timestampInSegment := segment.GetStartPosition().GetTimestamp()
+				timestampInTarget := channel.GetSeekPosition().GetTimestamp()
```

Comment:
for growing segment `gs`, if it's not in the unflushed segments list of current target, and the timestamp is between current and next, it will be released, which breaks the data consistency